apps/voting/vite.config.ts

```diff
@@ -27,43 +33,69 @@ function validateURL(envField: string): void {
-    throw new Error(
-        `Unexpected value for environment variable "CCD_ELECTION_NETWORK": ${process.env.CCD_ELECTION_NETWORK} (should be either "testnet" or "mainnet")`,
-    );
-}
+function getConfig(): Config {
+    // Validate network
+    if (![undefined, 'mainnet', 'testnet'].includes(process.env.CCD_ELECTION_NETWORK)) {
+        throw new Error(
+            `Unexpected value for environment variable "CCD_ELECTION_NETWORK": ${process.env.CCD_ELECTION_NETWORK} (should be either "testnet" or "mainnet")`,
+        );
+    }
 
-// Validate contract address
-if (!process.env.CCD_ELECTION_CONTRACT_ADDRESS?.match(/<\d*,\d*>/)) {
-    throw new Error('Environment variable "CCD_ELECTION_CONTRACT_ADDRESS" must be specified in the format "<1234,0>"');
-}
+    const [, index, subindex] =
+        process.env.CCD_ELECTION_CONTRACT_ADDRESS?.match(/<(\d*),(\d*)>/) ??
+        (() => {
+            throw new Error(
+                'Environment variable "CCD_ELECTION_CONTRACT_ADDRESS" must be specified in the format "<1234,0>"',
+            );
+        })();
 
-// Validate node URL
-validateURL('CCD_ELECTION_NODE');
+    // Validate node URL
+    validateURL('CCD_ELECTION_NODE');
 
-// Validate backend API URL
-validateURL('CCD_ELECTION_BACKEND_API');
+    return {
+        node: process.env.CCD_ELECTION_NODE ?? DEFAULT_NODE,
+        contractAddress: { index, subindex },
+        network: process.env.CCD_ELECTION_NETWORK ?? DEFAULT_NETWORK,
+    };
+}
 
-// https://vitejs.dev/config/
-export default defineConfig({
+const viteConfig: UserConfig = {
     plugins: [
         react({ plugins: [['@swc-jotai/react-refresh', {}]] }),
         tsconfigPaths(),
-        wasm(),
+        wasm() as PluginOption,
         topLevelAwait(), // For legacy browser compatibility
     ],
+    worker: {
+        plugins: [topLevelAwait(), wasm() as PluginOption],
+    },
     define: {
-        'process.env': {
-            CCD_ELECTION_NETWORK: process.env.CCD_ELECTION_NETWORK ?? DEFAULT_NETWORK,
-            CCD_ELECTION_NODE: process.env.CCD_ELECTION_NODE ?? DEFAULT_NODE,
-            CCD_ELECTION_CONTRACT_ADDRESS: process.env.CCD_ELECTION_CONTRACT_ADDRESS,
-            CCD_ELECTION_BACKEND_API: process.env.CCD_ELECTION_BACKEND_API ?? DEFAULT_BACKEND_API,
-        },
+        DEVICE_NAME: JSON.stringify(uuid()),
+        BACKEND_API: JSON.stringify(DEFAULT_BACKEND_API),
     },
     resolve: {
         alias: {
             '@concordium/rust-bindings': '@concordium/rust-bindings/bundler', // Resolve bundler-specific wasm entrypoints.
         },
     },
+};
+
+// https://vitejs.dev/config/
+export default defineConfig(({ command }) => {
+    if (command === 'serve') {
+        // Mimic the configuration injection from the backend
+        const config = getConfig();
+        viteConfig.plugins!.push(
+            handlebars({
+                context: { config: JSON.stringify(config) },
```

Comment:
I was wondering about this in the rust code as well, but why do we make a string here for the `config` value? Can't it just be the `config` object itself?

Reply:
It has to be a string as the value is injected in the template through string interpolation. If it is not a JSON string, the value ends up in the template as

```html
<script>
  const config = [Object object];
</script>
```

Which isn't of much use to anyone 😄

Reply:
And it's the same (ish) with the rust counterpart